src/modules/universal-discovery/universal.discovery.module.js

```diff
@@ -250,22 +286,53 @@ export default class UniversalDiscoveryModule extends Component {
     }
 
     /**
-     * Returns true or false if locations bookmark data was cached
-     * if not, checks it and then caches it
+     * Removes or adds bookmark depending on if it exists or not
+     *
+     * @method toggleBookmark
+     * @param {Object} location
+     * @memberof UniversalDiscoveryModule
+     */
+    toggleBookmark(location) {
+        const locationId = location.id;
+        const isBookmarked = this.isBookmarked(locationId);
+
+        if (this.isBookmarkLoading(locationId)) {
+            return;
+        }
+
+        this.setBookmarkLoadingStatus(locationId, true);
+
+        const { restInfo } = this.props;
+        const toggleBookmark = isBookmarked ? removeBookmark : addBookmark;
+        const onBookmarkToggled = isBookmarked ? this.onBookmarkRemoved : this.onBookmarkAdded;
+        const bookmarkToggled = new Promise((resolve) => toggleBookmark(restInfo, locationId, resolve));
+
+        bookmarkToggled
+            .then(() => {
+                onBookmarkToggled(location);
+                this.setBookmarkLoadingStatus(locationId, false);
+            })
+            .catch(showErrorNotification);
+    }
+
+    /**
+     * Returns
      *
      * @method isBookmarked
      * @param {String} locationId
+     * @returns {Boolean}
      * @memberof UniversalDiscoveryModule
      */
     isBookmarked(locationId) {
         const { bookmarked } = this.state;
-        if (locationId in bookmarked) {
-            return bookmarked[locationId];
-        }
+        const locationBookmarkChecked = locationId in bookmarked;
 
-        this.checkIsBookmarked(locationId);
+        if (!locationBookmarkChecked) {
+            this.checkIsBookmarked(locationId);
```

Comment:
Now I realize we have method `isBookmarked` and `checkIsBookmarked` little bit confusing.

Reply:
What do you suggest instead?

Reply:
`fetchBookmarkedStatus`?